ee/control/consumers/notificationconsumer/notify_consumer.go

```diff
@@ -0,0 +1,261 @@
+package notificationconsumer
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	desktopRunner "github.com/kolide/launcher/ee/desktop/runner"
+	"github.com/kolide/launcher/pkg/osquery"
+	"go.etcd.io/bbolt"
+)
+
+// Consumes notifications from control server, tracks when notifications are sent to end user
+type NotificationConsumer struct {
+	db                          *bbolt.DB
+	runner                      notifier
+	logger                      log.Logger
+	notificationRetentionPeriod time.Duration
+	cleanupInterval             time.Duration
+	ctx                         context.Context
+	cancel                      context.CancelFunc
+}
+
+// The desktop runner fullfils this interface -- it exists primarily for testing purposes.
+type notifier interface {
+	SendNotification(title, body string) error
+}
+
+// Represents notification received from control server; SentAt is set by this consumer after sending.
+// For the time being, notifications are per-end user device and not per-user.
+type notification struct {
+	Title      string    `json:"title"`
+	Body       string    `json:"body"`
+	ID         string    `json:"id"`
+	ValidUntil string    `json:"valid_until"` // ISO8601 format
+	SentAt     time.Time `json:"sent_at,omitempty"`
+}
+
+const (
+	// Identifier for this consumer.
+	NotificationSubsystem = "desktop_notifier"
+
+	// ISO 8601 -- used for valid_until field
+	iso8601Format = "2006-01-02T15:04:05-0700"
+
+	// Approximately 6 months
+	defaultRetentionPeriod = time.Hour * 24 * 30 * 6
+
+	// How frequently to check for old notifications
+	defaultCleanupInterval = time.Hour * 12
+)
+
+type notificationConsumerOption func(*NotificationConsumer)
+
+func WithLogger(logger log.Logger) notificationConsumerOption {
+	return func(nc *NotificationConsumer) {
+		nc.logger = log.With(logger,
+			"component", NotificationSubsystem,
+		)
+	}
+}
+
+func WithNotificationRetentionPeriod(ttl time.Duration) notificationConsumerOption {
+	return func(nc *NotificationConsumer) {
+		nc.notificationRetentionPeriod = ttl
+	}
+}
+
+func WithCleanupInterval(cleanupInterval time.Duration) notificationConsumerOption {
+	return func(nc *NotificationConsumer) {
+		nc.cleanupInterval = cleanupInterval
+	}
+}
+
+func NewNotifyConsumer(db *bbolt.DB, runner *desktopRunner.DesktopUsersProcessesRunner, ctx context.Context, opts ...notificationConsumerOption) (*NotificationConsumer, error) {
```

Comment:
No, it needs to be a `desktopRunner.DesktopUsersProcessesRunner` -- the consumer sends notification requests to the desktop user processes (where they're then handled by a `notifier`) via the runner, because the runner owns those processes.

Reply:
I've added a diagram to the ADR to hopefully document this a little better -- https://github.com/kolide/launcher/blob/959fa225dcb28b200a0ed763f14526a9c479f2c5/docs/architecture/notifications.md#lifecycle-of-a-notification

Reply:
Oh, might not have looked hard enough. I thought that's what the `notifier` interface was for?

Reply:
The `notifier` interface exists at this point only so I can mock the desktop runner in notification consumer tests.

I can rename it to something else -- it's named `notifier` and not something more specific mostly by virtue of having refactored this like 4 separate times. 😅

Reply:
Seems a little weird, but okay. if it's only used in the tests, it could be declared in the tests.

Reply:
Renamed it to `userProcessesRunner` and updated the comment for accuracy